mempool/mempool.go

```diff
@@ -270,10 +304,19 @@ func (mem *Mempool) filterTxs(blockTxsMap map[string]struct{}) []types.Tx {
 	goodTxs := make([]types.Tx, 0, mem.txs.Len())
 	for e := mem.txs.Front(); e != nil; e = e.Next() {
 		memTx := e.Value.(*mempoolTx)
+		// Remove the tx if its alredy in a block.
 		if _, ok := blockTxsMap[string(memTx.tx)]; ok {
-			// Remove the tx since already in block.
+			// remove from clist
 			mem.txs.Remove(e)
 			e.DetachPrev()
+
+			// remove from mempool cache
+			// we only enforce "at-least once" semantics and
+			// leave it to the application to implement "only-once"
+			// via eg. sequence numbers, utxos, etc.
+			// NOTE: expects caller of filterTxs to hold the lock
+			// (so we can't use mem.removeTxFromCacheMap)
+			delete(mem.cacheMap, string(memTx.tx))
```

Comment:
The cacheMap is pre-allocated, that's fine, but if we don't remove committed txs it will always be full, so we'll be doing extra type assertions and map look ups to remove something on every new transaction.

What happens if a tx is included in a block but the AppendTx fails? This could actually be a security vulnerability since if the tx pays feed in the app an attacker/validator could potentially replay the invalid appendtx many times and drain the account.  So this is either an argument for apps to ensure they enforce strict only-once semantics even when txs fail, or to formalize a fee mechanism in tendermint core to prevent that kind of attack.